File: hr/contests/lambda-calculi-sept14/bangalore-bank.hs
-- I wish memoization were easier in Haskell ....

cost :: Int -> Int -> [Int] -> Int
cost left right [] = 0
cost left right (digit:digits) = 1 + min
    ((abs (left  - digit)) + cost digit right digits)
    ((abs (right - digit)) + cost left  digit digits)

total_cost :: [Int] -> Int
total_cost digits = minimum [cost left right digits | left <- [1..10], right <- [1..10]]

main =
    getLine >>
    getLine >>= \ line ->
        putStrLn $ show $ total_cost [
            if digit == 0
                then 10
                else digit
            | digit <- [(read digit :: Int) | digit <- words line]]

File: hr/contests/lambda-calculi-sept14/bangalore-bank.scala
object Solution {
    val memo = new scala.collection.mutable.HashMap[(Int,Int,Int), Int]()
    var digits : Vector[Int] = null
    
    def cost(left : Int, right : Int, level : Int) : Int = {
        val key = (left, right, level)
        if (!memo.contains(key)) {
            if (level == digits.length) {
                memo.put(key, 0)
            }
            else {
                val digit = digits(level)
                memo.put(key, 1 + scala.math.min(
                    (left  - digit).abs + cost(digit, right, level+1),
                    (right - digit).abs + cost(left,  digit, level+1)))
            }
        }
        
        return memo.getOrElse(key, 0)
    }
    
    def main(args : Array[String]) : Unit = {
        scala.io.StdIn.readLine
        digits = scala.io.StdIn.readLine.
            split("\\s+").
            map(_.toInt).
            map((digit) => if (digit == 0) 10 else digit).
            toVector
        
        println((for (left <- 1 to 10; right <- 1 to 10) yield cost(left, right, 0)).min)
    }
}
